File: skills/openclix-init/templates/flutter/core/clix.dart
import '../engine/trigger_service.dart';
import '../models/clix_types.dart';
import '../services/config_loader.dart';
import '../services/config_validator.dart';
import '../services/utils.dart';

class ClixDependencies {
  final ClixLocalMessageScheduler messageScheduler;
  final CampaignStateRepositoryPort campaignStateRepository;
  final ClixClock? clock;
  final ClixLifecycleStateReader? lifecycleStateReader;
  final ClixLogger? logger;

  const ClixDependencies({
    required this.messageScheduler,
    required this.campaignStateRepository,
    this.clock,
    this.lifecycleStateReader,
    this.logger,
  });
}

class DefaultClock implements ClixClock {
  @override
  String now() => DateTime.now().toUtc().toIso8601String();
}

class DefaultLifecycleStateReader implements ClixLifecycleStateReader {
  String currentAppState = 'foreground';

  @override
  String getAppState() => currentAppState;

  @override
  void setAppState(String state) {
    currentAppState = state;
  }

  @override
  void dispose() {}
}

class DefaultLogger implements ClixLogger {
  ClixLogLevel logLevel;

  static const Map<ClixLogLevel, int> logLevelOrder = {
    ClixLogLevel.debug: 0,
    ClixLogLevel.info: 1,
    ClixLogLevel.warn: 2,
    ClixLogLevel.error: 3,
    ClixLogLevel.none: 4,
  };

  DefaultLogger(this.logLevel);

  @override
  void setLogLevel(ClixLogLevel level) {
    logLevel = level;
  }

  @override
  void debug(String message, [Object? argument]) {
    if (shouldLog(ClixLogLevel.debug)) {
      print(
        '[OpenClix] DEBUG: $message${argument != null ? ' $argument' : ''}',
      );
    }
  }

  @override
  void info(String message, [Object? argument]) {
    if (shouldLog(ClixLogLevel.info)) {
      print('[OpenClix] INFO: $message${argument != null ? ' $argument' : ''}');
    }
  }

  @override
  void warn(String message, [Object? argument]) {
    if (shouldLog(ClixLogLevel.warn)) {
      print('[OpenClix] WARN: $message${argument != null ? ' $argument' : ''}');
    }
  }

  @override
  void error(String message, [Object? argument]) {
    if (shouldLog(ClixLogLevel.error)) {
      print(
        '[OpenClix] ERROR: $message${argument != null ? ' $argument' : ''}',
      );
    }
  }

  bool shouldLog(ClixLogLevel targetLevel) {
    return logLevelOrder[targetLevel]! >= logLevelOrder[logLevel]!;
  }
}

bool isRemoteEndpoint(String endpoint) {
  return endpoint.startsWith('http://') || endpoint.startsWith('https://');
}

class Clix {
  static ClixConfig? config;
  static TriggerService? triggerService;
  static bool initialized = false;
  static CampaignStateRepositoryPort? campaignStateRepository;
  static ClixLocalMessageScheduler? messageScheduler;
  static ClixClock? clock;
  static ClixLifecycleStateReader? lifecycleStateReader;
  static ClixLogger? logger;
  static ClixDependencies? dependencies;

  Clix._();

  static Future<void> initialize(
    ClixConfig config,
    ClixDependencies dependencies,
  ) async {
    if (initialized) {
      throw StateError(
        'Clix is already initialized. Call Clix.reset() before re-initializing.',
      );
    }

    Clix.config = config;
    Clix.dependencies = dependencies;
    messageScheduler = dependencies.messageScheduler;
    campaignStateRepository = dependencies.campaignStateRepository;
    clock = dependencies.clock ?? DefaultClock();
    lifecycleStateReader =
        dependencies.lifecycleStateReader ?? DefaultLifecycleStateReader();
    logger = dependencies.logger ?? DefaultLogger(config.logLevel);
    logger?.setLogLevel(config.logLevel);

    triggerService = TriggerService(createTriggerServiceDependencies());

    logger?.info('Initializing OpenClix SDK...');

    if (isRemoteEndpoint(config.endpoint)) {
      try {
        final requestHeaders = <String, String>{
          ...(config.extraHeaders ?? const <String, String>{}),
        };

        if (config.projectId != null && config.projectId!.isNotEmpty) {
          requestHeaders['x-openclix-project-id'] = config.projectId!;
        }

        if (config.apiKey != null && config.apiKey!.isNotEmpty) {
          requestHeaders['x-openclix-api-key'] = config.apiKey!;
        }

        final loadedConfig = await loadConfig(
          config.endpoint,
          options: ConfigLoaderOptions(
            headers: requestHeaders,
            timeoutMs: config.sessionTimeoutMs,
          ),
        );

        final validationResult = validateConfig(loadedConfig);

        if (validationResult.valid) {
          for (final warning in validationResult.warnings) {
            logger?.warn(
              'Config validation warning [${warning.code}]: ${warning.message}',
            );
          }

          triggerService?.replaceConfig(loadedConfig);

          try {
            await evaluateInternal('app_boot');
          } catch (evaluationError) {
            logger?.warn(
              'Initial app_boot evaluation failed:',
              evaluationError,
            );
          }

          logger?.info(
            'Config loaded successfully '
            '(version: ${loadedConfig.configVersion}, '
            'campaigns: ${loadedConfig.campaigns.length})',
          );
        } else {
          for (final error in validationResult.errors) {
            logger?.error(
              'Config validation error [${error.code}]: ${error.message}',
            );
          }
          logger?.warn(
            'Config validation failed. SDK initialized without campaign config.',
          );
        }
      } catch (loadError) {
        logger?.warn(
          'Failed to load config from endpoint. SDK initialized without campaign config. '
          'Use ClixCampaignManager.replaceConfig() to set config manually.',
          loadError,
        );
      }
    } else {
      logger?.info(
        'Non-HTTP endpoint provided. '
        'Use ClixCampaignManager.replaceConfig() to set campaign config.',
      );
    }

    initialized = true;
    logger?.info('OpenClix SDK initialized successfully.');
  }

  static Future<void> trackEvent(
    String name, [
    Map<String, JsonValue>? properties,
  ]) async {
    assertInitialized();

    final event = Event(
      id: generateUUID(),
      name: name,
      sourceType: EventSourceType.app,
      properties: properties,
      createdAt: clock!.now(),
    );

    logger?.debug('Event tracked (not persisted): $name');

    try {
      await evaluateInternal('event_tracked', event);
    } catch (evaluationError) {
      logger?.warn("Evaluation after event '$name' failed:", evaluationError);
    }
  }

  static Future<void> reset() async {
    final loggerAtResetStart = logger;

    if (campaignStateRepository != null) {
      try {
        await campaignStateRepository!.clearCampaignState();
      } catch (error) {
        loggerAtResetStart?.warn(
          'Failed to clear campaign state during reset:',
          error,
        );
      }
    }

    if (messageScheduler != null) {
      try {
        final pendingMessages = await messageScheduler!.listPending();
        for (final pendingMessage in pendingMessages) {
          await messageScheduler!.cancel(pendingMessage.id);
        }
      } catch (error) {
        loggerAtResetStart?.warn(
          'Failed to clear scheduled messages during reset:',
          error,
        );
      }
    }

    try {
      lifecycleStateReader?.dispose();
    } catch (error) {
      loggerAtResetStart?.warn(
        'Failed to dispose lifecycle state reader during reset:',
        error,
      );
    }

    config = null;
    triggerService = null;
    initialized = false;
    campaignStateRepository = null;
    messageScheduler = null;
    clock = null;
    lifecycleStateReader = null;
    logger = null;
    dependencies = null;

    loggerAtResetStart?.info('OpenClix SDK reset complete.');
  }

  static void setLogLevel(ClixLogLevel level) {
    logger?.setLogLevel(level);
  }

  static void handleAppForeground() {
    if (!initialized) return;

    lifecycleStateReader?.setAppState('foreground');
    logger?.debug('App entered foreground');

    evaluateInternal('app_foreground').catchError((Object error) {
      logger?.warn('app_foreground evaluation failed:', error);
      return null;
    });
  }

  static TriggerService? getTriggerServiceInternal() => triggerService;

  static ClixClock? getClockInternal() => clock;

  static ClixLogger? getLoggerInternal() => logger;

  static CampaignStateRepositoryPort? getCampaignStateRepositoryInternal() {
    return campaignStateRepository;
  }

  static ClixLocalMessageScheduler? getMessageSchedulerInternal() {
    return messageScheduler;
  }

  static bool isInitializedInternal() => initialized;

  static void assertInitialized() {
    if (!initialized) {
      throw StateError(
        'Clix is not initialized. Call Clix.initialize() before using the SDK.',
      );
    }
  }

  static TriggerServiceDependencies createTriggerServiceDependencies() {
    return TriggerServiceDependencies(
      campaignStateRepository: campaignStateRepository!,
      messageScheduler: messageScheduler!,
      clock: clock!,
      logger: logger!,
    );
  }

  static Future<TriggerResult?> evaluateInternal(
    String trigger, [
    Event? event,
  ]) async {
    if (triggerService == null) {
      return null;
    }

    return triggerService!.trigger(
      TriggerContext(trigger: trigger, event: event, now: clock?.now()),
    );
  }
}

File: skills/openclix-init/templates/android/core/Clix.kt
package ai.openclix.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import ai.openclix.engine.TriggerService
import ai.openclix.engine.TriggerServiceDependencies
import ai.openclix.models.CampaignStateRepository
import ai.openclix.models.ClixClock
import ai.openclix.models.ClixConfig
import ai.openclix.models.ClixLifecycleStateReader
import ai.openclix.models.ClixLocalMessageScheduler
import ai.openclix.models.ClixLogLevel
import ai.openclix.models.ClixLogger
import ai.openclix.models.Event
import ai.openclix.models.EventSourceType
import ai.openclix.models.TriggerContext
import ai.openclix.models.TriggerResult
import ai.openclix.services.loadConfig
import ai.openclix.services.validateConfig
import java.util.UUID

private const val DEFAULT_CONFIG_TIMEOUT_MILLISECONDS = 10_000

object Clix {

    private var config: ClixConfig? = null
    private var triggerService: TriggerService? = null

    @Volatile
    private var initialized: Boolean = false

    private var campaignStateRepository: CampaignStateRepository? = null
    private var messageScheduler: ClixLocalMessageScheduler? = null
    private var clock: ClixClock? = null
    private var lifecycleStateReader: MutableLifecycleStateReader? = null
    private var logger: DefaultLogger? = null

    @JvmStatic
    suspend fun initialize(
        config: ClixConfig,
        campaignStateRepository: CampaignStateRepository,
        messageScheduler: ClixLocalMessageScheduler,
        clock: ClixClock? = null,
        logger: ClixLogger? = null
    ) {
        if (initialized) {
            throw IllegalStateException(
                "Clix is already initialized. Call Clix.reset() before re-initializing."
            )
        }

        this.config = config
        this.campaignStateRepository = campaignStateRepository
        this.messageScheduler = messageScheduler
        this.clock = clock ?: DefaultClock()

        val defaultLogger = if (logger != null) {
            DefaultLogger(config.logLevel).also { createdLogger ->
                createdLogger.delegate = logger
            }
        } else {
            DefaultLogger(config.logLevel)
        }
        this.logger = defaultLogger

        val defaultLifecycleStateReader = MutableLifecycleStateReader()
        lifecycleStateReader = defaultLifecycleStateReader

        triggerService = TriggerService(createTriggerServiceDependencies())

        defaultLogger.info("Initializing OpenClix SDK...")

        val endpoint = config.endpoint
        val isRemoteEndpoint = endpoint.startsWith("http://") || endpoint.startsWith("https://")

        if (isRemoteEndpoint) {
            try {
                val requestHeaders = mutableMapOf<String, String>()
                config.extraHeaders?.let { requestHeaders.putAll(it) }
                config.projectId?.let { requestHeaders["x-openclix-project-id"] = it }
                config.apiKey?.let { requestHeaders["x-openclix-api-key"] = it }

                val loadedConfig = loadConfig(
                    endpoint = endpoint,
                    extraHeaders = if (requestHeaders.isNotEmpty()) requestHeaders else null,
                    timeoutMs = config.sessionTimeoutMs ?: DEFAULT_CONFIG_TIMEOUT_MILLISECONDS
                )

                if (loadedConfig != null) {
                    val validationResult = validateConfig(loadedConfig)

                    if (validationResult.valid) {
                        for (warning in validationResult.warnings) {
                            defaultLogger.warn("Config validation warning [${warning.code}]: ${warning.message}")
                        }

                        triggerService?.replaceConfig(loadedConfig)

                        try {
                            evaluate("app_boot")
                        } catch (evaluationError: Exception) {
                            defaultLogger.warn(
                                "Initial app_boot evaluation failed:",
                                evaluationError.message ?: evaluationError.toString()
                            )
                        }

                        defaultLogger.info(
                            "Config loaded successfully (version: ${loadedConfig.config_version}, campaigns: ${loadedConfig.campaigns.size})"
                        )
                    } else {
                        for (error in validationResult.errors) {
                            defaultLogger.error("Config validation error [${error.code}]: ${error.message}")
                        }
                        defaultLogger.warn("Config validation failed. SDK initialized without campaign config.")
                    }
                }
            } catch (loadError: Exception) {
                defaultLogger.warn(
                    "Failed to load config from endpoint. SDK initialized without campaign config. " +
                            "Use ClixCampaignManager.replaceConfig() to set config manually.",
                    loadError.message ?: loadError.toString()
                )
            }
        } else {
            defaultLogger.info(
                "Non-HTTP endpoint provided. Use ClixCampaignManager.replaceConfig() to set campaign config."
            )
        }

        initialized = true
        defaultLogger.info("OpenClix SDK initialized successfully.")
    }

    @JvmStatic
    suspend fun trackEvent(name: String, properties: Map<String, Any?>? = null) {
        assertInitialized()

        val event = Event(
            id = UUID.randomUUID().toString(),
            name = name,
            source_type = EventSourceType.APP,
            properties = properties,
            created_at = clock!!.now()
        )

        logger?.debug("Event tracked (not persisted): $name")

        try {
            evaluate("event_tracked", event)
        } catch (evaluationError: Exception) {
            logger?.warn(
                "Evaluation after event '$name' failed:",
                evaluationError.message ?: evaluationError.toString()
            )
        }
    }

    @JvmStatic
    suspend fun reset() {
        val activeLogger = logger

        campaignStateRepository?.let { repository ->
            try {
                repository.clearCampaignState()
            } catch (error: Exception) {
                activeLogger?.warn(
                    "Failed to clear campaign state during reset:",
                    error.message ?: error.toString()
                )
            }
        }

        messageScheduler?.let { scheduler ->
            try {
                val pendingMessages = scheduler.listPending()
                for (pendingMessage in pendingMessages) {
                    scheduler.cancel(pendingMessage.id)
                }
            } catch (error: Exception) {
                activeLogger?.warn(
                    "Failed to clear scheduled messages during reset:",
                    error.message ?: error.toString()
                )
            }
        }

        config = null
        triggerService = null
        initialized = false
        campaignStateRepository = null
        messageScheduler = null
        clock = null
        lifecycleStateReader = null
        logger = null

        activeLogger?.info("OpenClix SDK reset complete.")
    }

    @JvmStatic
    fun setLogLevel(level: ClixLogLevel) {
        logger?.level = level
    }

    @JvmStatic
    fun handleAppForeground() {
        if (!initialized) return

        lifecycleStateReader?.setState("foreground")
        logger?.debug("App entered foreground")

        @Suppress("OPT_IN_USAGE")
        GlobalScope.launch(Dispatchers.Default) {
            try {
                evaluate("app_foreground")
            } catch (evaluationError: Exception) {
                logger?.warn(
                    "app_foreground evaluation failed:",
                    evaluationError.message ?: evaluationError.toString()
                )
            }
        }
    }

    internal fun getTriggerServiceInternal(): TriggerService? = triggerService

    internal fun getClockInternal(): ClixClock? = clock

    internal fun getLoggerInternal(): ClixLogger? = logger

    internal fun getCampaignStateRepositoryInternal(): CampaignStateRepository? = campaignStateRepository

    internal fun getMessageSchedulerInternal(): ClixLocalMessageScheduler? = messageScheduler

    internal fun isInitializedInternal(): Boolean = initialized

    private fun assertInitialized() {
        if (!initialized) {
            throw IllegalStateException(
                "Clix is not initialized. Call Clix.initialize() before using the SDK."
            )
        }
    }

    private fun createTriggerServiceDependencies(): TriggerServiceDependencies {
        return TriggerServiceDependencies(
            campaignStateRepository = campaignStateRepository!!,
            scheduler = messageScheduler!!,
            clock = clock!!,
            logger = logger!!
        )
    }

    internal suspend fun evaluate(
        trigger: String,
        event: Event? = null
    ): TriggerResult? {
        val activeTriggerService = triggerService ?: return null
        return activeTriggerService.trigger(
            TriggerContext(
                trigger = trigger,
                event = event,
                now = clock?.now()
            )
        )
    }
}

private class DefaultClock : ClixClock {
    override fun now(): String {
        val formatter = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US)
        formatter.timeZone = java.util.TimeZone.getTimeZone("UTC")
        return formatter.format(java.util.Date())
    }
}

internal class MutableLifecycleStateReader : ClixLifecycleStateReader {
    @Volatile
    private var appState: String = "foreground"

    override fun getAppState(): String = appState

    fun setState(nextState: String) {
        appState = nextState
    }
}

internal class DefaultLogger(
    @Volatile var level: ClixLogLevel
) : ClixLogger {
    var delegate: ClixLogger? = null

    private val logTag = "OpenClix"

    override fun debug(msg: String, vararg args: Any?) {
        if (level.priority <= ClixLogLevel.DEBUG.priority) {
            delegate?.debug(msg, *args) ?: android.util.Log.d(logTag, formatMessage(msg, args))
        }
    }

    override fun info(msg: String, vararg args: Any?) {
        if (level.priority <= ClixLogLevel.INFO.priority) {
            delegate?.info(msg, *args) ?: android.util.Log.i(logTag, formatMessage(msg, args))
        }
    }

    override fun warn(msg: String, vararg args: Any?) {
        if (level.priority <= ClixLogLevel.WARN.priority) {
            delegate?.warn(msg, *args) ?: android.util.Log.w(logTag, formatMessage(msg, args))
        }
    }

    override fun error(msg: String, vararg args: Any?) {
        if (level.priority <= ClixLogLevel.ERROR.priority) {
            delegate?.error(msg, *args) ?: android.util.Log.e(logTag, formatMessage(msg, args))
        }
    }

    private fun formatMessage(message: String, args: Array<out Any?>): String {
        return if (args.isEmpty()) message else "$message ${args.joinToString(" ")}"
    }
}
